runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,412 @@
+import org.apache.beam.sdk.transforms.reflect.DoFnSignatures;
+import org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.GlobalWindow;
+import org.apache.beam.sdk.transforms.windowing.OutputTimeFns;
+import org.apache.beam.sdk.transforms.windowing.PaneInfo;
+import org.apache.beam.sdk.util.KeyedWorkItem;
+import org.apache.beam.sdk.util.TimeDomain;
+import org.apache.beam.sdk.util.TimerInternals;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.util.state.StateNamespace;
+import org.apache.beam.sdk.util.state.StateNamespaces;
+import org.apache.beam.sdk.util.state.StateTag;
+import org.apache.beam.sdk.util.state.StateTags;
+import org.apache.beam.sdk.util.state.ValueState;
+import org.apache.beam.sdk.util.state.WatermarkHoldState;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.TupleTag;
+import org.joda.time.Instant;
+
+/**
+ * A utility transform that executes a <a
+ * href="https://s.apache.org/splittable-do-fn">splittable</a> {@link DoFn} by expanding it into a
+ * network of simpler transforms: pair with restriction, split restrictions, assign unique key and
+ * group by it (in order to get access to per-key state and timers), then process.
+ *
+ * <p>This transform is intended as a helper for internal use by runners when implementing {@code
+ * ParDo.of(splittable DoFn)}, but not for direct use by pipeline writers.
+ */
+@Experimental(Experimental.Kind.SPLITTABLE_DO_FN)
+public class SplittableParDo<
+        InputT, OutputT, RestrictionT, TrackerT extends RestrictionTracker<RestrictionT>>
+    extends PTransform<PCollection<InputT>, PCollection<OutputT>> {
+  private final DoFn<InputT, OutputT> fn;
+
+  /**
+   * Creates the transform for the given original {@link ParDo} and {@link DoFn}.
+   *
+   * @param fn The splittable {@link DoFn} inside the original {@link ParDo} transform.
+   */
+  public SplittableParDo(DoFn<InputT, OutputT> fn) {
+    this.fn = fn;
+  }
+
+  @Override
+  public PCollection<OutputT> apply(PCollection<InputT> input) {
+    PCollection.IsBounded isFnBounded =
+        DoFnSignatures.INSTANCE.getOrParseSignature(fn.getClass()).isBounded();
+    Coder<RestrictionT> restrictionCoder =
+        DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn).invokeGetRestrictionCoder();
+    Coder<ElementRestriction<InputT, RestrictionT>> splitCoder =
+        ElementRestrictionCoder.of(input.getCoder(), restrictionCoder);
+
+    return input
+        .apply(
+            "Pair with initial restriction",
+            ParDo.of(new PairWithRestrictionFn<InputT, OutputT, RestrictionT>(fn)))
+        .setCoder(splitCoder)
+        .apply("Split restriction", ParDo.of(new SplitRestrictionFn<InputT, RestrictionT>(fn)))
+        .setCoder(splitCoder)
```

Comment:
nit: This call to setCoder may not be necessary since the input type and output type are the same it should be propagated. That said, likely no harm in setting it.